# INF 8215 - Intelligence artif.: méthodes et algorithmes 
## Automne 2018 - TP1 - Méthodes de recherche 
### Membres de l'équipe
    - Fabrice Dugas (1690694)
    - Félix Boulet (1788287)
    - Yujia Ding (1801923)



## LE VÉLO À MONTRÉAL
Chaque année, Montréal accueille à peu près 10 millions de touristes. Soucieuse de la qualité de leur séjour, Tourisme Montréal a entamé un projet de développement d’une nouvelle application mobile afin d’assister les touristes lors de leurs déplacements dans la ville. Cette application a pour but d’aider l’utilisateur à planifier sa visite des importantes attractions de la ville, de la façon la plus efficace possible (ie, sur la durée la plus courte). Étant donné qu’il a été observé que le moyen de transport privilégié des touristes pour explorer Montréal est le vélo, cette application a pour but de générer des circuits cyclables de durée minimale. Plus précisément, étant donné une liste d’attractions munie de points de départ et d’arrivée, la tâche est de proposer, à chaque fois, un chemin qui passe par toutes les attractions indiquées une seule fois, qui débute au point de départ et qui s’achève au point d’arrivée et dont la durée de trajet est minimale.

<img src="images/montreal.png" alt="" width="800"/>

Le travail demandé dans ce TP est de développer l’algorithme interne de l’application. Nous explorerons trois mécanismes de résolution différents :
1. Définition et exploration naïve d’un arbre de recherche
2. Exploration plus efficace en utilisant l’algorithme A*
3. Optimisation locale en utilisant une métaheuristique de recherche à voisinage variable (Variable Neighborhood Search, VNS)

## PRÉSENTATION DU PROBLÈME
Une façon naturelle de représenter notre problème est d’utiliser un graphe $G=(V, A)$ dirigé et complet. Chaque sommet dans $V$ est une attraction donnée et chaque arc dans $A$ représente une piste cyclable entre deux attractions distinctes. Chaque paire de sommets $i$ et $j$ est reliée par une paire d’arcs $a_{ij}$ et $a_{ji}$ dont les poids respectifs $w(a_{ij})$ et $w(a_{ji})$ ne sont pas nécessairement égaux. Concrètement, ces poids représentent la durée du trajet d’un sommet à l’autre (ainsi, $w$ est telle que $w : A \to\mathbb R^+$).

La liste des attractions à visiter est indiquée comme la suite $P = (p_1, ..., p_m)$ où $p_1$ et $p_m$ sont les sommets de départ et d’arrivée, respectivement.

## 1. DÉFINITION ET EXPLORATION NAÏVE D’UN ARBRE DE RECHERCHE (5 points)
Définissons un arbre de recherche $\mathcal{T}$ où chaque nœud représente une solution partielle $S$. Soient $V(S) \subseteq V$ et $A(S) \subset A$ l’ensemble des sommets visités et l’ensemble des arêtes sélectionnées, respectivement. Ainsi, le coût d’une solution est donné par :
$$g(S) = \sum_{a \in A(S)} w(a)$$

Seule l’origine est visitée initialement. Ainsi, la racine de l’arbre de recherche contient une solution partielle vide $S_{\textrm{root}}$ telle que $V(S_{\textrm{root}})=\{p_1\}$ et $A(S_{\textrm{root}}) = \emptyset$.

<img src="images/tree1.png" alt="" width="100"/>

À la suite de cela, les nœuds subséquents dans l’arbre sont tous créés en ajoutant, à chaque solution partielle $S$, un sommet subséquent dans $P\backslash V(S)$ avec l’arc correspondant dans $A$ qui relie ce sommet à la dernière attraction visitée. Le sommet $p_m$ n’est ajouté qu’à la fin, lorsqu’il est le seul sommet non encore visité. Plus formellement, si on note le sommet à ajouter $c$ et le dernier sommet visité $c'$, alors la nouvelle solution partielle obtenue est $V(S) \gets V(S) \cup \{c\}$ et $A(S) \gets A(S) \cup \{(c’,c)\}$.

Ci-dessous est un exemple de l’arbre étendu depuis sa racine où $c'$ = $p_1$ :

<img src="images/tree2.png" alt="" width="400"/>

À la fin, les feuilles de l’arbre sont des solutions complètes :

<img src="images/tree3.png" alt="" width="600"/>

### 1.1 Code
La fonction fournie ci-dessous permet d’extraire d’un fichier un graphe qui répond aux spécifications détaillées plus haut. Cette fonction retourne une $\texttt{ndarray}$ ($\texttt{graph}$) de taille $|V|\times |V|$ où $\texttt{graph[i,j]}$ représente le temps nécessaire pour traverser la piste cyclable de $i$ vers $j$.



In [2]:
import numpy as np

def read_graph():
    return np.loadtxt("montreal", dtype='i', delimiter=',')

graph = read_graph()

Notre première tâche est de définir la classe qui représente une solution partielle. Son constructeur est donné et reçoit comme argument la liste des sommets (attractions $P$) à visiter et le graphe ($G$). Celui-ci crée la solution $S_{\textrm{root}}$ avec les attributs suivants :
- $\texttt{g}$ : le coût de la solution partielle
- $\texttt{visited}$ : représente $V(S)$, discuté plus haut. Par définition, $\mathtt{vistited[-1]}$ représente le dernier sommet ajouté, $ c $.
- $\texttt{not}\_\texttt{visited}$ : représente $P\backslash V(S)$
- $\texttt{graph}$: représente le graphe G

Ensuite, il est demandé d’implanter la méthode $\texttt{add}$ qui mets à jour la solution partielle en ajoutant une nouvelle attraction à visiter parmi la liste $\texttt{not}\_\texttt{visited}$. Cette méthode reçoit comme arguments l’index du sommet à visiter parmi $\texttt{not}\_\texttt{visited}$.

Implantez $\texttt{add}$ :

In [3]:
import copy

class Solution:
    def __init__(self, places, graph):
        """
        places: a list containing the indices of attractions to visit
        p1 = places[0]
        pm = places[-1]
        """
        self.g = 0 # current cost
        self.h = 0 # estimate
        self.graph = graph 
        self.visited = [places[0]] # list of already visited attractions
        self.not_visited = copy.deepcopy(places[1:]) # list of attractions not yet visited
        
    def add(self, idx):
        """
        Adds the point in position idx of not_visited list to the solution
        """
        ret = copy.deepcopy(self)
        end = ret.not_visited.pop(idx)
        begin = ret.visited[-1]
        cost = ret.graph[begin, end]
        ret.g += cost
        ret.visited.append(end)
        return ret
    
    def __𝚕𝚝__(𝚜𝚎𝚕𝚏, 𝚘𝚝𝚑𝚎𝚛):
        return (self.g + self.h) < (other.g + other.h)
    
    def swap(self, i, j):
        """
        Swaps two visited nodes at index i and j
        i and j cannot be the first or last element in visited
        """
        if i == 0 or i == len(self.visited)-1 or j == 0 or j == len(self.visited)-1:
            print("!!! WARNING: SWAPPING BEGIN OR END NODES !!!")
        
        ret = copy.deepcopy(self)
        
        # actually swap the nodes
        tmp = ret.visited[i]
        ret.visited[i] = ret.visited[j]
        ret.visited[j] = tmp
        
        # calculate new cost
        # could be more efficient by adding to the current cost but YOLO
        self.g = 0
        for idx, place in enumerate(ret.visited):
            if idx == 0:
                continue
            first = ret.visited[idx-1]
            self.g += self.graph[first, place]
        
        return ret
        
        

La prochaine étape est d’implanter une stratégie de parcours de l’arbre de recherche. Une première méthode simple est naïve est de mettre en œuvre une recherche en largeur ([Breadth-first search](https://moodle.polymtl.ca/pluginfile.php/444662/mod_resource/content/1/recherche_en_largeur.mp4), BFS).

Implantez $\texttt{bfs}$ qui mets en œuvre cette recherche. Elle prend en arguments le graphe courant ainsi que la liste des attractions à visiter $P$ et elle retourne la meilleure solution trouvée.

In [3]:
from queue import Queue

def bfs(graph, places):
    """
    Returns the best solution which spans over all attractions indicated in 'places'
    """
    explored_node_count = 0
    best_cost = 100000000000
    frontier = Queue()
    solution = Solution(places, graph)
    best_solution = None
    frontier.put(solution)
    while frontier.qsize() > 0:
        solution = frontier.get()
        if len(solution.not_visited) == 0 and solution.g < best_cost:
            best_cost = solution.g
            best_solution = solution
        for idx, place in enumerate(solution.not_visited):
            if place != places[-1] or len(solution.not_visited) == 1:
                explored_node_count += 1
                frontier.put(solution.add(idx))
                
    return best_solution, explored_node_count

### 1.2 Expérimentations

On propose trois exemples d’illustration pour tester notre recherche en largeur. Le premier exemple prend en compte 7 attractions, le second 10 et le dernier 11. Vu que cette recherche énumère toutes les solutions possibles, le troisième exemple risque de prendre un temps considérable à s’achever.

Mettez en œuvre ces expériences et notez le nombre de nœuds explorés ainsi que le temps de calcul requis.

In [4]:
import time 

#test 1  --------------  OPT. SOL. = 27
start_time = time.time()
places=[0, 5, 13, 16, 6, 9, 4]
sol, explored = bfs(graph=graph, places=places)
print(sol.g)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

NameError: name 'bfs' is not defined

In [46]:
#test 2 -------------- OPT. SOL. = 30
start_time = time.time()
places=[0, 1, 4, 9, 20, 18, 16, 5, 13, 19]
sol, explored = bfs(graph=graph, places=places)
print(sol.g)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

30
149920
--- 4.647337913513184 seconds ---


In [47]:
#test 3 -------------- OPT. SOL. = 26
start_time = time.time()
places=[0, 2, 7, 13, 11, 16, 15, 7, 9, 8, 4]
sol, explored = bfs(graph=graph, places=places)
print(sol.g)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

26
1349289
--- 43.338595390319824 seconds ---


| Test | Nb. noeuds | Temps (s) |
|------|------------|-----------|
| 1    | 445        | 0.0150063 |
| 2    | 149920     | 4.0327137 |
| 3    | 1349289    | 39.531402 |

## 2. RECHERCHE GUIDÉE À L’AIDE DE L’ALGORITHME A\* (7.5 points)
Pour notre deuxième méthode de recherche, au lieu d’énumérer toutes les solutions possibles, nous effectuons une recherche guidée à l’aide de l’algorithme A\*. Comme vu en classe, A\* est une recherche où les nœuds à explorer sont priorisés en fonction du coût courant d’une solution $g(S)$ ainsi que d’une estimation du coût restant vers la solution finale donné par une heuristique $h(S)$.

Dans le cas d’une minimisation, $h(S)$ est une borne inférieure du coût réel restant et on priorise l’exploration des nœuds dont $f(S) = g(S)+h(S)$ est le plus petit. Avec cette méthode, la première solution complète trouvée est assurément la solution optimale.

Pour une solution donnée $S$ avec un dernier sommet visité $c$, une possible fonction $h$ est telle que :

$h(S) =$ Le poids du chemin le plus court entre $c$ et $p_m$ dans le sous graphe $G_S$ contenant les sommets $P\backslash V(S) \cup \{c\}$

Remarque que ce chemin le plus court utilisé dans le calcul de l’estimation $h$ entre l’attraction courante et l’arrivée ne passera pas nécessairement pas tous les sommets restants.


Notre algorithme A\* se présente comme ceci :
1. Définir l’arbre de recherche $\mathcal{T}$ exactement comme auparavant. Le calcul de $h$ pour la solution initiale est inutile : c’est la seule solution qu’on a.
2. Sélectionner le meilleur nœud candidat pour expansion. La solution partielle $S_b$ de ce nœud candidat est telle que :

   $$ f(S_b) \leq f(S) \quad \forall S \in \mathcal{T} \qquad S_B, S \text{ pas encore sélectionnés}$$

   Si $S_b$ est une solution complète, l’algorithme s’arrête et $S_b$ est assurément la solution optimale, sinon on continue à l’étape 3.
3. Créer des solutions subséquentes qui connectent la dernière attraction visitée à chacune des attractions restantes. Attention, on ignore l’arrivée tant que celle-ci n’est pas la seule qui reste.
 - Mettez à jour les listes des sommets visités et non visités
 - Calculez $g$ et $h$ pour chaque solution
 - Insérer la nouvelle solution partielle dans l’arbre.
4. Répéter 2 et 3.


### 2.1 Code
Commençons d’abord par compléter la classe $\texttt{Solution}$ pour prendre en compte les changements nécessaires à A\* (on a besoin notamment d’un attribut supplémentaire pour l’estimation $h$).

On verra plus tard que A\* s’implante à l’aide d’une file de priorité (priority queue). Pour que celle-ci marche, il est nécessaire de surcharger (overload) l’opérateur de comparaison « < » relatif à nos objets $\texttt{Solution}$. En sachant ce qui fait qu'une solution est meilleure qu’une autre pour l'exploration, implanter la méthode $\_\_\texttt{lt}\_\_$ dans $\texttt{Solution}$. Son prototype est $\_\_\texttt{lt}\_\_\texttt{(self, other)}$.

Maintenant, nous devons implanter la fonction d’estimation $h$. Pour cela, on utilise l’[algorithme de Dijkstra](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm) pour trouver le chemin le plus court entre la dernière attraction visitée $c$ et l’arrivée $p_m$. Il est possible d’adapter cet algorithme pour qu’il s’arrête dès que le chemin le plus court entre c et pm est trouvé.

**Prescriptions d’implantation :**
- Appliquer Dijkstra pour trouver le chemin le plus court entre $c$ et $p_m$
- Retourner le poids de ce chemin

In [48]:
import heapq

def fastest_path_estimation(sol):
    """
    Returns the time spent on the fastest path between 
    the current vertex c and the ending vertex pm
    """
    if not sol.not_visited:
        return 0
    
    c = sol.visited[-1]
    pm = sol.not_visited[-1]
    
    # Array of tuples of the form (cost, node)
    q = [(0, c)]
    
    # Map the array to a set of tuples of the form (None, x) where x is the node id.
    remaining = set(map(lambda x: (None, x), sol.not_visited))
    processed = set()
    
    # While there are nodes in the priority queue
    while q:
        (cost, node) = heapq.heappop(q) # Pop the first one
        if node not in processed: # If the node wasn't processed already (at a better cost/priority)
            if node == pm: # If the node is the target, we're done
                return cost
            processed.add(node) # If not, add it to the processed nodes
            if node in remaining:
                remaining.remove(node) # Should always be true, but just in case
            for previous_cost, neighbor in remaining: # Check all remaining neighbors that have no shortest path yet
                cost_now  = cost + sol.graph[node, neighbor]
                if previous_cost is None or cost_now < previous_cost: # If it has a better cost, change it and push to the queue
                    previous_cost = cost_now
                    heapq.heappush(q, (cost_now, neighbor))
    return None # path is unreachable
    

Finalement, il est temps d’implanter A\*. On aura besoin d’une file de priorité qui retournera toujours le meilleur nœud candidat de $\mathcal{T}$ pour l’étendre (l’opérateur surchargé de comparaison assure cela).

**Prescriptions d’implantation (cf. détail des étapes de l’algorithme plus haut) :**
- Tant que les solutions extraites de la file de priorité ne sont pas complètes :
  *	Sélectionner et étendre le nœud extrait de la file comme détaillé plus haut
  * Calculer $g$ et $h$ pour chaque nouvelle solution partielle obtenue
  * Remettre ces solutions dans la file
- Retourner la première solution complète extraite de la file (c’est la solution optimale)

In [49]:
import heapq

def A_star(graph, places):
    """
    Performs the A* algorithm
    """
    
    explored_count = 0

    # blank solution
    root = Solution(graph=graph, places=places)

    # search tree T
    T = []
    heapq.heapify(T)
    heapq.heappush(T, root)

    # While there are nodes in the priority queue
    while T:
        solution = heapq.heappop(T) # Pop the first one
        if len(solution.not_visited) == 0:
            return solution, explored_count
        for idx, place in enumerate(solution.not_visited):
            if place != places[-1] or len(solution.not_visited) == 1:
                explored_count += 1
                temp_sol = solution.add(idx)
                temp_sol.h = fastest_path_estimation(temp_sol)
                heapq.heappush(T, temp_sol)
    return None, explored_count
    

### 2.2 Expérimentations

On ajoute un Quatrième exemple d’exécution avec 15 attractions. Là encore, mettez en œuvre ces expériences avec le nouvel algorithme A\* conçu et notez le nombre de nœuds explorés ainsi que le temps de calcul requis.

In [9]:
import time

#test 1  --------------  OPT. SOL. = 27
start_time = time.time()
places=[0, 5, 13, 16, 6, 9, 4]
astar_sol, explored = A_star(graph=graph, places=places)
print(astar_sol.g)
print(astar_sol.visited)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

NameError: name 'A_star' is not defined

In [51]:
#test 2  --------------  OPT. SOL. = 30
start_time = time.time()
places=[0, 1, 4, 9, 20, 18, 16, 5, 13, 19]
astar_sol, explored = A_star(graph=graph, places=places)
print(astar_sol.g)
print(astar_sol.visited)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

30
[0, 1, 4, 5, 9, 13, 16, 18, 20, 19]
1302
--- 0.0746614933013916 seconds ---


In [52]:
#test 3  --------------  OPT. SOL. = 26
start_time = time.time()
places=[0, 2, 7, 13, 11, 16, 15, 7, 9, 8, 4]
astar_sol, explored = A_star(graph=graph, places=places)
print(astar_sol.g)
print(astar_sol.visited)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

26
[0, 2, 7, 7, 9, 13, 15, 16, 11, 8, 4]
3831
--- 0.21531248092651367 seconds ---


In [53]:
#test 4  --------------  OPT. SOL. = 40
start_time = time.time()
places=[0, 2, 20, 3, 18, 12, 13, 5, 11, 16, 15, 4, 9, 14, 1]
astar_sol, explored = A_star(graph=graph, places=places)
print(astar_sol.g)
print(astar_sol.visited)
print(explored)
print("--- %s seconds ---" % (time.time() - start_time))

40
[0, 3, 5, 13, 15, 18, 20, 16, 11, 12, 14, 9, 4, 2, 1]
624491
--- 51.97768020629883 seconds ---


#### Heuristique Dijkstra

| Test | Nb. noeuds | Temps (s) |
|------|------------|-----------|
| 1    | 120        | 0.0049860 |
| 2    | 1302       | 0.0540085 |
| 3    | 3831       | 0.1389732 |
| 4    | 624491     | 38.987563 |

#### Heuristique Minimum Spanning Arborescence


### 2.3 Une meilleure borne inférieure

Notre algorithme A\* est déjà beaucoup plus efficace qu’une recherche naïve. Cependant, la qualité de l’heuristique $h$ a un très grand impact sur la vitesse de A\*. Une heuristique plus serrée devrait accélérer A\* de façon significative. Notre estimation $h$ basée sur Dijkstra est très large à cause du fait qu’elle ne considère pas toutes les attractions restantes.

Une meilleure heuristique pourrait être basée sur la **Spanning Arborescence of Minimum Weight** qui s’apparente à une Minimum Spanning Tree pour graphes orientés. On propose de construire une telle Spanning Arborescence sur le reste des attractions $P\backslash V(S) \cup \{c\}$. Ici la racine est la dernière attraction visitée $c$. Une façon classique de résoudre ce problème est d’utiliser l’[algorithme de Edmonds](https://en.wikipedia.org/wiki/Edmonds%27_algorithm).

Implantez cet algorithme et refaites les expériences avec A\* en utilisant cette nouvelle heuristique :

In [7]:
# Inspired by https://github.com/mlbright/edmonds/blob/master/edmonds/edmonds.py

def reverse(edge_dict):
    reverse_dict = {}
    for node in edge_dict:
        for (next_node, cost) in edge_dict[node].items():
            if next_node in reverse_dict:
                reverse_dict[next_node][node] = cost
            else:
                reverse_dict[next_node] = {node: cost}
    return reverse_dict
        
def find_cycle(node, edges, visited, cycle):
    visited.add(node) # We visit the provided node
    cycle += [node] # Add it to the current cycle
    
    if node not in edges: # If the node is not in the edge dict, meaning it has no minimum neighbor...
        return cycle # Return the current cycle
    
    for next_node in edges[node]: # For each minimum neighbor of the current node (there should only be one)
        if next_node not in visited: # If we didn't visit the neighbor yet,
            cycle = find_cycle(next_node, edges, visited, cycle) # Check for a longer cycle
    return cycle # When we have no next_node to check (because it's visited), that means we're at the end of the cycle

def merge_cycles(cycle, minimum_edges, reverse_edges, sol):
    all_in_edges = []
    min_internal = None
    min_internal_weight = 100000000000
    
    for node in cycle:
        for idx, edge in enumerate(graph[:, node]):
            if idx != node and idx in sol.not_visited:
                if idx in cycle:
                    if min_internal is None or edge < min_internal_weight:
                        min_internal = (node, idx)
                        min_internal_weight = edge
                else:
                    all_in_edges.append((node, idx))
    
    min_external = None
    min_modified_weight = 100000000000
    
    print(reverse_edges)
    for (dest, src) in all_in_edges:
        ndest, ncost = reverse_edges[dest].popitem()
        reverse_edges[dest][ndest] = ncost
        weight = graph[dest, src] - (ncost - min_internal_weight)
        if min_external is None or weight < min_modified_weight:
            min_external = (dest, src)
            min_modified_weight = weight
    
    ndest, ncost = reverse_edges[min_external[0]].popitem()
    rem = (min_external[0], ndest)
    reverse_edges[min_external[0]].clear()
    
    if min_external[1] in reverse_edges:
        reverse_edges[min_external[1]][min_external[0]] = ncost
    else:
        reverse_edges[min_external[1]] = {min_external[0] : ncost}
        
    if rem[1] in minimum_edges:
        if rem[0] in minimum_edges[rem[1]]:
            del minimum_edges[rem[1]][rem[0]]
    
    if min_external[1] in minimum_edges:
        minimum_edges[min_external[1]][min_external[0]] = ncost
    else:
        minimum_edges[min_external[1]] = {min_external[0] : ncost}   

def minimum_spanning_arborescence(sol):
    root = sol.visited[-1]
    minimum_edges = {}
    
    # for each node that's not been visited yet,
    for node in sol.not_visited:
        min_edge_val = 100000000000
        min_edge_idx = -1
        
        # for each edge that goes to the node,
        for idx, edge in enumerate(graph[:, node]):
            # if the edge has less cost than the best one found yet, save it since it's a closer neighbor
            if idx != node and idx in sol.not_visited and edge < min_edge_val:
                min_edge_val = edge
                min_edge_idx = idx
                
        # Add to the dict the provenance of the min edge to the node
        if node in minimum_edges:
            minimum_edges[node][min_edge_idx] = min_edge_val
        else:
            minimum_edges[node] = {min_edge_idx : min_edge_val}
    
    # At this point we should have a dict containing the "nearest neighbor" of each node
    cycles = []
    visited = set()
    
    # for each node in the dict,
    for node in minimum_edges:
        # if the node has not been visited as part of cycle detection yet
        if node not in visited:
            # check for cycles that include this node
            cycles.append(find_cycle(node, minimum_edges, visited, []))
    
    reverse_edges = reverse(minimum_edges)
    print(reverse_edges)
    print(cycles)
    for cycle in cycles:
        merge_cycles(cycle, minimum_edges, reverse_edges, sol)
    
    h = 0
    for node in minimum_edges:
        for (src, cost) in minimum_edges[node]:
            h += cost
        
    return h
places = range(21)
print(minimum_spanning_arborescence(Solution(graph=graph, places=places)))


{2: {1: 4, 3: 2}, 3: {2: 2, 7: 1}, 5: {4: 1}, 9: {5: 1, 10: 1, 17: 1}, 8: {6: 1}, 4: {8: 1}, 6: {9: 1, 12: 1, 13: 1, 14: 1, 15: 1}, 14: {11: 1}, 11: {16: 2}, 15: {18: 1}, 16: {19: 2}, 19: {20: 2}}
[[1, 2, 3], [4, 5, 9, 6, 8], [7], [10], [11, 14], [12], [13], [15], [16], [17], [18], [19], [20]]
{2: {1: 4, 3: 2}, 3: {2: 2, 7: 1}, 5: {4: 1}, 9: {5: 1, 10: 1, 17: 1}, 8: {6: 1}, 4: {8: 1}, 6: {9: 1, 12: 1, 13: 1, 14: 1, 15: 1}, 14: {11: 1}, 11: {16: 2}, 15: {18: 1}, 16: {19: 2}, 19: {20: 2}}


KeyError: 1

## 3. RECHERCHE LOCALE À VOISINAGE VARIABLE  (7.5 points)

Cette fois-ci, au lieu de construire une solution optimale depuis une solution vide, on commence d’une solution complète, non-optimale, qu’on améliore à l’aide d’une recherche locale en utilisant une recherche locale à voisinage variable ([Variable Neighborhood Search](https://en.wikipedia.org/wiki/Variable_neighborhood_search), VNS).

<img src="images/vns.png" alt="" width="800"/>

### 3.1 Code

On commence par créer une solution initiale. Celle-ci est une suite ordonnée des attractions de $p_1$ à $p_m$ dans $P$. Pour cela, on fait appel à une [recherche en profondeur (Depth-First Search, DFS)](https://moodle.polymtl.ca/pluginfile.php/445484/mod_resource/content/1/recherche_en_profondeur.mp4) qu’on arrête aussitôt qu’une solution complète est trouvée. Pour aider à diversifier la recherche, la méthode permettant de générer une solution initiale peut être randomisée de telle sorte que l'algorithme VNS puisse lancer la recherche dans différentes régions de l'espace solution. Ainsi, dans la fonction DFS, la sélection de l'enfant pour continuer la recherche doit être aléatoire.

**Prescriptions d’implantation :**
- Mettre en œuvre une recherche en profondeur
- Créer un objet $\texttt{Solution}$ relatif à cette solution
- Ajuster les attributs de cet objet avec les bonnes valeurs de coûts et d’attractions visitées
- Retourner la solution trouvée.

In [4]:
from random import shuffle, randint

def initial_sol(graph, places):
    """
    Return a completed initial solution
    """
    return dfs(graph, places)
    

def dfs(graph, places):
    """
    Performs a Depth-First Search
    """    
    solution = Solution(places, graph)
    while len(solution.not_visited) > 0:
        not_visited = solution.not_visited
        if len(solution.not_visited) != 1:
            # on enleve la destination des choix pour la prochaine solution a visiter
            not_visited = solution.not_visited[0:-1]
            
        index = randint(0, len(not_visited)-1)
        solution = solution.add(index)
        
    return solution
    

Pour définir une VNS, il faut définir les $k_\textrm{max}$ voisinages de recherche locale possibles. Pour notre problème, une bonne et simple répartition des voisinages est telle qu’un voisinage $k$ correspond à la permutation de $k$-paires de sommets dans $V(S)$.

On appelle **shaking** l’étape de génération d’une solution dans le voisinage $k$. Le travail qui suit correspond à l’implantation de cette étape. $\texttt{shaking}$ admet 3 arguments que sont la solution de départ, l’indice du voisinage $k$ ainsi que le graph courant.

Attention, avant d’implanter $\texttt{shaking}$, il est nécessaire de créer une méthode $\texttt{swap}$ dans la classe $\texttt{Solution}$. Cette méthode permet de mettre en œuvre la permutation dans une solution donnée (en mettant à jour tous les attributs nécessaires pour que la solution soit cohérente).

**Prescriptions d’implantation de shaking :**
- Sélectionner au hasard deux indices $i$ et $j$ différents et tels que $i, j \in \{2,...,m-1\}$
- Faire une copie de la solution courante et faire la permutation
- Retourner la solution créée

In [5]:
def shaking(sol, k):
    """
    Returns a solution on the k-th neighrboohood of sol
    """
    print("shaking")
    solution = sol
    for derp in range(k):
        
        print(derp)
        
        i = randint(2,len(solution.visited)-1)
        j = randint(2,len(solution.visited)-1)
        while i == j:
            # assure que i sera different de j
            j = randint(2,len(solution.visited)-1)
        
        solution = sol.swap(i,j)
        
    return solution

Une dernière étape essentielle dans une VNS est l’application d’un algorithme de recherche locale à la solution issue du shaking. Pour cela, on propose la recherche locale 2-opt. Celle-ci intervertit deux arcs dans la solution, à la recherche d’une qui est meilleure.

Pour un sommet $ i $, soit $ i '$ le successeur immédiat de $ i $ dans la séquence $ V (S) $. L'algorithme 2-opt fonctionne comme suit: pour chaque paire de sommets non consécutifs $ i, j $, vérifiez si en échangeant la position des sommets $ i '$ et $ j $ entraîne une amélioration du coût de la solution. Si oui, effectuez cet échange. Ce processus se répète jusqu'à ce qu'il n'y ait plus d'échanges rentables. On réalise cette opération pour toutes les paires d’arcs éligibles à la recherche du plus petit coût.

<img src="images/2opt.png" alt="" width="800"/>

<img src="images/2opt2.png" alt="" width="800"/>


Implantez $\texttt{local}\_\texttt{search}\_\texttt{2opt}$. 

**Prescriptions d’implantation :**
- Considérer chaque paire d’indices $i = \{2,..,m-3\}$ and $j = \{i+2, m-1\}$
- Si l’échange donne un plus bas coût, on le réalise
- Répéter jusqu’à optimum local.

In [6]:
def local_search_2opt(sol):
    """
    Apply 2-opt local search over sol
    """
    best_solution = sol
    
    found_better_solution = True
    while found_better_solution:
        found_better_solution = False
        solution = best_solution
        for i in range(2,len(solution.visited)-3):
            for j in range(i+2,len(solution.visited)-1):
                neighbour = solution.swap(i,j)
                if neighbour.g < best_solution.g:
                    best_solution = neighbour
                    found_better_solution = True
    
    return best_solution


Finalement, il est temps d'implanter notre VNS. La méthode $\texttt{vns}$ reçoit une solution complète, le graphe courant, le nombre maximal de voisinages et un temps de calcul limite. Celle-ci retourne la solution optimale trouvée

**Prescriptions d’implantation :**
- À chaque itération, la VNS génère une solution dans le k-ème voisinage (shaking) à partir de la meilleure solution courante et applique une recherche locale 2-opt dessus
- Si la nouvelle solution trouvée a un meilleur coût, mettre à jour la meilleure solution courante
- Répéter le processus jusqu'à $\texttt{t}\_\texttt{max}$

In [7]:
def vns(sol, k_max, t_max):
    """
    Performs the VNS algorithm
    """
    k = 1
    solution = sol           # solution courante
    best_solution = sol      # meilleure solution a vie
    start_time = time.time()
    current_time = start_time
    while current_time < start_time + t_max:
        print("Debut {}".format(solution.g))
        
        solution = shaking(solution, k)
        solution = local_search_2opt(solution)
        
        print(solution.g)
        
        if solution.g < best_solution.g:
            best_solution = solution
    
        # go to next neighborhood
        k = k % k_max + 1
        current_time = time.time()
    
    return best_solution

### 3.2 Experiments

Mettez en oeuvre la VNS sur les exemples d'illustration suivants et raportez les solutions obtenue:

In [10]:
# test 1  --------------  OPT. SOL. = 27
places=[0, 5, 13, 16, 6, 9, 4]
sol = initial_sol(graph=graph, places=places)
start_time = time.time()
vns_sol = vns(sol=sol, k_max=10, t_max=1)
print(vns_sol.g)
print(vns_sol.visited)
print("--- %s seconds ---" % (time.time() - start_time))

Debut 40
shaking
0
!!! WARNING: SWAPPING BEGIN OR END NODES !!!


KeyboardInterrupt: 

In [ ]:
#test 2  --------------  OPT. SOL. = 30
places=[0, 1, 4, 9, 20, 18, 16, 5, 13, 19]
sol = initial_sol(graph=graph, places=places)

start_time = time.time()
vns_sol = vns(sol=sol, k_max=10, t_max=1)
print(vns_sol.g)
print(vns_sol.visited)

print("--- %s seconds ---" % (time.time() - start_time))

In [ ]:
# test 3  --------------  OPT. SOL. = 26
places=[0, 2, 7, 13, 11, 16, 15, 7, 9, 8, 4]
sol = initial_sol(graph=graph, places=places)

start_time = time.time()
vns_sol = vns(sol=sol, k_max=10, t_max=1)
print(vns_sol.g)
print(vns_sol.visited)
print("--- %s seconds ---" % (time.time() - start_time))

In [ ]:
# test 4  --------------  OPT. SOL. = 40
places=[0, 2, 20, 3, 18, 12, 13, 5, 11, 16, 15, 4, 9, 14, 1]
sol = initial_sol(graph=graph, places=places)

start_time = time.time()
vns_sol = vns(sol=sol, k_max=10, t_max=1)
print(vns_sol.g)
print(vns_sol.visited)
print("--- %s seconds ---" % (time.time() - start_time))

## 4. BONUS (1 point)

Expliquez dans quelle situation chacun des algorithmes développés est plus approprié (prenez en compte l’évolutivité du problème)

## 5 Date de remise
Tout devra être remis avant le 6 Octobre à 23h55. Tout travail en retard sera pénalisé d'une valeur de 10% par jour de retard.